### PR TITLE
Capsule-box collision

### LINKED ISSUE #77: Capsule <> Box
Capsule <> box collision function.

### PR DESCRIPTION
**Pull request by camevor**
Capsule-box collision based on mjc

includes / depends on #129 and resolves #77

### CODE REVIEW

File: mujoco_warp/_src/collision_primitive.py [hunk 1]
```diff
@@ -23,6 +23,9 @@
 from .types import GeomType
 from .types import Model
 
+_MINVAL = 1e-15
```
Comment (adenzler-nvidia) [inline mujoco_warp/_src/collision_primitive.py]: is this different to types.MJ_MINVAL?

**Review by adenzler-nvidia — COMMENTED** [review]
1 comment, but deferring the full review to someone who knows the ins-and-outs of the C implementation. @yuvaltassa probably?

File: mujoco_warp/_src/collision_primitive.py (hunk 1 above)
Comment (camevor) [reply mujoco_warp/_src/collision_primitive.py]: It's the same, thanks for spotting this!

File: mujoco_warp/_src/collision_primitive.py [hunk 2]
```diff
@@ -19,10 +19,13 @@
 from .math import closest_segment_to_segment_points
 from .math import make_frame
 from .math import normalize_with_norm
+from .types import MJ_MINVAL
 from .types import Data
 from .types import GeomType
 from .types import Model
 
+_TINY_VAL = 1e-6
```
Comment (thowell) [inline mujoco_warp/_src/collision_primitive.py:27]: can this variable be replaced with `MJ_MINVAL`?

File: mujoco_warp/_src/collision_primitive.py [hunk 3]
```diff
@@ -860,6 +861,402 @@ def sphere_box(
...
+
+  bestdistmax = margin + 2.0 * (
+    cap.size[0] + cap.size[1] + box.size[0] + box.size[1] + box.size[2]
+  )
+
+  # keep track of closest point
+  bestdist = wp.float32(bestdistmax)
+  bestsegmentpos = wp.float32(-12)
+  cltype = wp.int32(-4)
+  clface = wp.int32(-12)
```
Comment (adenzler-nvidia) [inline mujoco_warp/_src/collision_primitive.py:942]: are these magic numbers taken from mujoco? Maybe add some comments about what they mean?

File: mujoco_warp/_src/collision_primitive.py [hunk 4]
```diff
@@ -860,6 +861,402 @@ def sphere_box(
...
+  dd = wp.vec2(halfaxis.x, halfaxis.y)
+  s = wp.vec2(box.size.x, box.size.y)
+  secondpos = wp.float32(-4.0)
+
+  l = wp.length_sq(dd)
+
+  uu = dd.x * s.y
+  vv = dd.y * s.x
+  # w = dd.x * p.y - dd.y * p.x
+  w_neg = dd.x * p.y - dd.y * p.x < 0
```
Comment (adenzler-nvidia) [inline mujoco_warp/_src/collision_primitive.py:1074]: is this comment still necessary?

File: mujoco_warp/_src/collision_primitive.py [hunk 5]
```diff
@@ -860,6 +861,402 @@ def sphere_box(
...
+          if 0 < e1 and e1 < secondpos:
+            secondpos = e1
+
+          e1 = (-box.size[i] - tmp1[i]) * ha_r
+          if 0 < e1 and e1 < secondpos:
+            secondpos = e1
+
+      secondpos *= wp.float32(mul)
+
+  # skip:
```
Comment (adenzler-nvidia) [inline mujoco_warp/_src/collision_primitive.py]: what does the skip comment refer to?

**Review by adenzler-nvidia — COMMENTED** [review]
some comments. This it pretty tricky code, I'm not sure I can find issues here. But generally looks good, and if it follows MJC it should be fine.

File: mujoco_warp/_src/collision_primitive.py (hunk 5 above)
Comment (erikfrey) [reply mujoco_warp/_src/collision_primitive.py]: Unfortunately: https://github.com/google-deepmind/mujoco/blob/main/src/engine/engine_collision_box.c#L425

This is my least favorite bit of C MuJoCo.  I have promised @kbayes untold riches the day we delete these gotos.

Cursory scan of this code looks like @camevor has correctly undone the gotos - agree might be worth deleting this comment or otherwise adding `# goto skip;` comments elsewhere to situate this comment.

File: mujoco_warp/_src/collision_driver_test.py [hunk 6]
```diff
@@ -225,6 +225,28 @@ class CollisionTest(parameterized.TestCase):
...
+          <worldbody>
+            <geom type="box" pos="0 0 0" size=".5 .4 .9" />
+            <body pos="0 0 1.5" >
+              <geom type="capsule" size="0.5 0.8"/>
+              <freejoint/>
+            </body>
+          </worldbody>
+        </mujoco>
+        """,
+    "capsule_box_edge": """
```
Comment (erikfrey) [inline mujoco_warp/_src/collision_driver_test.py]: Capsule box is complicated!  Is this also testing capsule on box corner?  If not, let's add it.  Generally speaking, please add a few more fixtures to fully exercise that kernel logic which is pretty complex.

**Review by erikfrey — APPROVED** [review]
Generally LGTM but let's add one or two more test fixtures to fully exercise this complex kernel.

File: mujoco_warp/_src/collision_primitive.py (hunk 3 above)
Comment (camevor) [reply mujoco_warp/_src/collision_primitive.py:942]: Yes, these are just like in the mjc implementation. I'll add some explanations

File: mujoco_warp/_src/collision_primitive.py (hunk 4 above)
Comment (camevor) [reply mujoco_warp/_src/collision_primitive.py:1074]: No, removed it

File: mujoco_warp/_src/collision_primitive.py (hunk 5 above)
Comment (camevor) [reply mujoco_warp/_src/collision_primitive.py]: I left this in to provide an anchor when comparing against the mjc implementation, but I agree that it's confusing more than anything else. I'll do a cleanup pass over all the comments.

File: mujoco_warp/_src/collision_driver_test.py (hunk 6 above)
Comment (camevor) [reply mujoco_warp/_src/collision_driver_test.py]: Agreed! I added a two more fixtures (capsule on corner and capsule flat on face -> requires 2 contacts) and more strict contact checking in the test.

**Comment by camevor** [comment]
Thanks for your reviews and comments! Added some more test fixtures and cleaned up the comments in the code.
